File: src/cmcandy/C_Language_Answers/_0020.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace ::std;

class Solution
{
public:
    bool isValid(string s)
    {   
        int len = s.size();
        stack<char> st;
        for (int i = 0; i < len; i++)
        {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[')
            {
                st.push(s[i]);
            }
            else if (s[i] == ')' || s[i] == '}' || s[i] == ']')
            {   
                
                if (st.empty()||(st.top() != s[i]-1&&st.top() != s[i]-2))
                {
                    return false;
                }
                else
                {
                    st.pop();
                }
            }
        }
        return st.empty();
    }
};

File: src/cmcandy/Python_language_Answers/_0020.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        map = {')': '(', '}': '{', ']': '['}

        for c in s:
            if c in [')', '}', ']']:
                ex = map[c]
                if len(stack) > 0 and ex == stack[-1]:
                    stack.pop()
                else:
                    stack.append(c)
            else:
                stack.append(c)
        return len(stack) == 0